### PR TITLE
add documentation for hooks

### LINKED ISSUE #5066: Improve renewal hook documentation
The documentation for renewal hooks is very bad. Considering how complicated these hooks are it is important that the documentation for them is clear.
Related issues are: 
#2364 about how to run separate pre/post hooks for different renewals
#4561 lack of a single --help output for hooks
#4564 lack of cohesive tutorial for using hooks
#4534 documentation doesn't mention how command line arguments persist
#4323 unclear that global hooks override per lineage
#4103 unclear that renew-hook isn't run on first issuance
#3147 we need to support hook scripts directories
https://github.com/certbot/certbot/issues/5935 when are directory hooks run
Additionally at various times we say that pre and post hooks are per lineage, but we also say they're sometimes global, and we don't suggest a way to specify hooks for only one lineage

### PR DESCRIPTION
**Pull request by SwartzCr**
PR to fix #5066

### CODE REVIEW

**Comment by SwartzCr** [comment]
A first draft at expanding the hooks documentation, and an attempt to give a guide for new users on how to correctly configure renewal hooks.
@pconrad-fb - feel free to leave me comments on clarity and form, since I know that you were planning on updating the structure of using.rst
@bmw feel free to tell me if I have said anything that is factually inaccurate.

File: docs/using.rst [hunk 1]
```diff
@@ -655,8 +612,106 @@ The following files are available:
...
+failed hook causing renewal failures will indirectly result in a
+non-zero exit code. Hooks will only be run if a certificate is due for
+renewal, so you can run the above command frequently via ``certbot renew`` 
+without unnecessarily stopping your webserver.
+
+Unique ``--pre-hook`` and ``--post-hook`` hooks run before and after every renewal
+attempt. If two certificates have identical pre and post hooks, they will be run
+only once. Meaning renewal for the two certificates would follow this form:
+pre-hook is run, renewal for cert 1 is run, renewal for cert 2 is run, post-hook
+is run.
```
Comment (bmw) [inline docs/using.rst]: I'd expand this paragraph and possibly combine it with part of the next paragraph. In particular:

> Unique ``--pre-hook`` and ``--post-hook`` hooks run before and after every renewal attempt.

Do all unique pre and post hooks in my renewal conf files run before every renewal attempt? No, it's just those that are attempted to be renewed during the current run. Personally, the way I'd structure this is basically walk through the relevant steps when `certbot renew` is run.

File: docs/using.rst [hunk 2]
```diff
@@ -655,8 +612,106 @@ The following files are available:
...
+since Certbot exits with a non-zero exit code when renewals fail, a
+failed hook causing renewal failures will indirectly result in a
+non-zero exit code. Hooks will only be run if a certificate is due for
+renewal, so you can run the above command frequently via ``certbot renew`` 
+without unnecessarily stopping your webserver.
+
+Unique ``--pre-hook`` and ``--post-hook`` hooks run before and after every renewal
+attempt. If two certificates have identical pre and post hooks, they will be run
+only once. Meaning renewal for the two certificates would follow this form:
+pre-hook is run, renewal for cert 1 is run, renewal for cert 2 is run, post-hook
```
[the remark this answers is not in the document]
Comment (bmw) [reply docs/using.rst]: nit: I'd rephrase this to something like:

"Your pre-hook is run, certificate 1 is renewed, certificate 2 is renewed, then your post-hook is run."

to avoid confusion with the now deprecated renew hook.

File: docs/using.rst [hunk 3]
```diff
@@ -655,8 +612,106 @@ The following files are available:
...
+is run.
+
+Note that certificates outside of the renewal window (30 days) will not have
```
Comment (bmw) [inline docs/using.rst]: Again, this is 30 day value is technically configurable, but you can probably leave this as is if you want.

File: docs/using.rst [hunk 4]
```diff
@@ -655,8 +612,106 @@ The following files are available:
...
+renewal attempted, and will therefore not cause pre and post hooks to be run.
+If you want your hook to run only after a successful renewal, use
+``--deploy-hook`` in a command like this.
```
Comment (bmw) [inline docs/using.rst]: It'd be good to expand the documentation on `--deploy-hook`. In particular, the set environment variables should be discussed but this doesn't have to happen in this PR.

File: docs/using.rst [hunk 5]
```diff
@@ -655,8 +612,106 @@ The following files are available:
...
+
+                   service some-daemon restart >/dev/null
+                   ;;
+           esac
+   done
+
+If you would like to change what hooks are being run during the renewal of a
+certificate you can either request that certificate again and specify all of
+the arguments you want to be associated with it (hooks and otherwise), or
+manually edit its renewal configuration file. For information on the renewal
```
Comment (bmw) [inline docs/using.rst]: A third option I thought of is to use:
```
certbot renew --force-renewal --cert-name <certname> <options you want to modify>
```
It's also worth noting that this isn't hook specific and some of this same stuff is discussed under "Renewing certificates" and "Modifying the Renewal Configuration File". Perhaps we could combine this into a single section in the future and just link to it here? Obviously, we don't have to do that now though.

File: docs/using.rst [hunk 6]
```diff
@@ -655,8 +612,106 @@ The following files are available:
...
+configuration file please refer to config-file_.
```
Comment (bmw) [inline docs/using.rst]: s/file/files

Also, this links to the section on INI configuration files which presumably isn't what you wanted. For your convenience, there's documentation about how to do this [here](http://www.sphinx-doc.org/en/stable/markup/inline.html#ref-role).

File: docs/using.rst [hunk 7]
```diff
@@ -655,8 +612,106 @@ The following files are available:
 
 .. _hooks:
 
-Pre and Post Validation Hooks
-=============================
+Hooks
+-----
```
Comment (bmw) [inline docs/using.rst]: Changing this from "=" to "-" made this a subsection of "Where are my certificates?". Is this what you wanted?

File: docs/using.rst [hunk 8]
```diff
@@ -655,8 +612,106 @@ The following files are available:
...
+the arguments you want to be associated with it (hooks and otherwise), or
+manually edit its renewal configuration file. For information on the renewal
+configuration file please refer to config-file_.
+
+More information about hooks can be found by running
+``certbot --help renew``.
+
+
+Manual Validation Hooks
+-----------------------
```
Comment (bmw) [inline docs/using.rst]: This is also a subsection of "Where are my certificates?"

File: docs/using.rst [hunk 9]
```diff
@@ -807,6 +862,15 @@ Certbot accepts a global configuration file that applies its options to all invo
...
 files that can be found in ``/etc/letsencrypt/renewal``.
 
+Renewal configuration files store all values specified when a certificate was first
+issued. This includes things like key size, domains, authentication and installation
+method, and pre/post/deploy hooks. It is not recommended to manually edit these values
+as incorrectly set values can cause renewal failure. If you would like to update the
+information set in a renewal configuration file it is recommended that you re-issue
+the certificate with the values that you would prefer. If you do choose to edit the
+configuration file directly it is recommended that you call ``certbot renew --dry-run``
+to ensure that you haven't broken that certificate's ability to renew.
```
Comment (bmw) [inline docs/using.rst]: Do we want this here? A lot of this information is found under "Modifying the Renewal Configuration File". This section heading can be renamed if you want.

File: docs/using.rst [hunk 10]
```diff
@@ -494,6 +438,17 @@ can run on a regular basis, like every week or every day). In that case,
...
+To test that the renewal of a cert will work you can use the ``--dry-run``
+command. Example:
+
+   certbot renew --dry-run
+   
+Which will attempt renewal of all of your certificates, regardless of their
+age, with test certificates from our staging server. Certificates tested this
+way will not save these test certificates, but if renewal is successful
+with ``--dry-run`` then you can have confidence that they will be successfully
+renewed when they are within 30 days of expiration when you call ``certbot renew``.
```
[the remark this answers is not in the document]
Comment (SwartzCr) [reply docs/using.rst]: I know, but it didn't seem pertinent to put that here

File: docs/using.rst [hunk 11]
```diff
@@ -650,11 +652,11 @@ non-zero exit code. Hooks will only be run if a certificate is due for
 renewal, so you can run the above command frequently via ``certbot renew`` 
 without unnecessarily stopping your webserver.
 
-Unique ``--pre-hook`` and ``--post-hook`` hooks run before and after every renewal
-attempt. If two certificates have identical pre and post hooks, they will be run
+Saved or specified hooks run before and after every renewal attempt.
```
Comment (bmw) [inline docs/using.rst]: I still think this sentence is a little misleading. You covered this same information earlier, so perhaps we should just delete it and adjust the next sentence to say something like:

"If two certificates are due for renewal and have identical pre and post hooks, when `certbot renew` is run, each hook will only be run once."

File: docs/using.rst (hunk 9 above)
Comment (bmw) [reply docs/using.rst]: I still think this section is out of place. Personally what I'd like to see is to have a new sentence added to the previous paragraph saying something like:

"To learn more about setting the values in this file see Modifying the Renewal Configuration File"

and then have this paragraph merged into that section.

Comment (bmw) [reply docs/using.rst]: These comments are unaddressed. Do you disagree with me or did you miss them?

**Comment by pconrad-fb** [comment]
Yes, will do. Hopefully tomorrow.

**Comment by pconrad-fb** [comment]
Absolutely. In fact, I wouldn't mind a quick sync-up. Do you and your team
have any time tomorrow?

Peter



Peter Conrad

Lead Technical Writer: Infrastructure | salesforce.com

Office: (415) 471-5265



<http://smart.salesforce.com/sig/pconrad//us_mb/default/link.html>

**Comment by sydneyli** [comment]
@pconrad-fb Sure! I'll shoot you an e-mail to set up a time.

**Comment by bmw** [comment]
What's the status of this PR? There are a lot of misconceptions about how our hooks work which this PR could help clear up.

File: docs/using.rst [hunk 12]
```diff
@@ -655,8 +612,113 @@ The following files are available:
...
+to ``stderr`` but renewal will be attempted anyway. A failing hook
+doesn't directly cause Certbot to exit with a non-zero exit code, but
+since Certbot exits with a non-zero exit code when renewals fail, a
+failed hook causing renewal failures will indirectly result in a
+non-zero exit code. Hooks will only be run if a certificate is due for
+renewal, so you can run the above command frequently via ``certbot renew`` 
+without unnecessarily stopping your webserver.
+
+If two certificates are due for renewal and have identical pre and post hooks, 
+when ``certbot reneww`` is run, each hook will only be run once.
```
Comment (pomegranited) [inline docs/using.rst]: nit: typo in ``certbot reneww``

**Comment by zoracon** [comment]
@bmw [when you get a moment] would it be helpful to take some of this PR and touch up the docs? Or close?
